File: src/main/java/lab/amateur/learning/leetcode/java/TwoSum.java
package lab.amateur.learning.leetcode.java;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    public static void main(String[] args) {
        int[] nums = {2, 7, 11, 15};
        int target = 9;
        System.out.println(Arrays.toString(twoSum(nums, target)));
    }

    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> d = new HashMap<>();
        for (int i = 0; ; ++i) {
            int x = nums[i];
            int y = target - x;
            if (d.containsKey(y)) {
                return new int[]{d.get(y), i};
            }
            d.put(x, i);
        }
    }
}

File: src/main/scala/lab/amateur/learning/leetcode/scala/TwoSum.scala
package lab.amateur.learning.leetcode.scala

object TwoSum {
  def main(args: Array[String]): Unit = {
    val nums = Array(2, 7, 11, 15)
    val target = 9
    println(twoSum(nums = nums, target = target).mkString("Array(", ", ", ")"))
  }

  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    val d = scala.collection.mutable.Map[Int, Int]()
    var ans: Array[Int] = Array()
    for (i <- nums.indices if ans.isEmpty) {
      val x = nums(i)
      val y = target - x
      if (d.contains(y)) {
        ans = Array(d(y), i)
      } else {
        d(x) = i
      }
    }
    ans
  }
}
